### ISSUE TITLE
[improvement]showing output of external programs (qaac)

### ISSUE THREAD

**Issue by mbassiouny33** [MScholtes/PS2EXE#97]
Hi

I have a script as simple as 

```
.\qaac.exe -V 94 .\aud.m4a 
```
QAAC (like FFmpeg) outputs on std err (the second channel) the progress report. With ffmpeg, I was able to find a workaround since the encoder offered it, but most other encoding tools are not as nice (x264.exe, qaac.exe, etc)
Here is the exe created with ps2exe
![image](https://user-images.githubusercontent.com/12175103/210182473-1a721dc4-c280-44d9-9607-bbea09a892b5.png)

and it remains messy till the end with 1 character per line, and here is the exe created with PowerShell steroids (another ps to exe tool):

[ezgif-1-b085be6f75.webm](https://user-images.githubusercontent.com/12175103/210182674-4c694954-0678-46df-93a2-e465e49ef00a.webm)

Are there any improvements that could be made to make ps2exe handle this properly?

**Comment by MScholtes** [comment]
Hello @mbassiouny33,

this obviously is a design flaw of _QAAC_, the error stream has to be used for error and not status messages. I don't know where the extra carriage returns beginning with 3.2% come from, but I must refer you to the authors of _QAAC_.

Maybe you can do something like
```powershell
.\qaac.exe -V 94 .\aud.m4a 2>&1
```
to redirect to the correct stream or omit the error output with **2>$NULL** ?

Greetings

Markus

**Comment by mbassiouny33** [comment]
Hi Markus

Thanks a lot for your reply and time

>this obviously is a design flaw of QAAC,

Indeed, I do realize it is not a conventional design, but for many audio/video encoding tools, it is very common, because the stdout is used to output the stream itself and std error is used for information/status messages. (as I mentioned few weeks ago in another issue)
FFmpeg, which is the well-known swiss army knife for encoders, behaves exactly like qaac, many other encoders and audio synthesizers act the same.

Redirecting with ` 2>&1` won't do neither :(  It will show 1 character per line and things will get messy like in the photo but without the "Error:" prefix.

The reason I suggested this improvement is because I see ISEsteroids handles the real-time progress report properly (as you can see in this video), so I assumed there is some room for improvement in ps2exe as well. Naturally, I understand you are under no obligation to accommodate the request for my particular use-case or to implement extra features but thought I would still report it.

Best

**Comment by MScholtes** [comment]
Hello @mbassiouny33,

**ISESteroids** uses a completely different process than **PS2EXE** to create executables.
**ISESteroids** creates a start-up EXE file, which then starts a hidden _Powershell_ with the "converted" script. The hidden _Powershell_ then behaves exactly like a normal _Powershell_ on input and output.
**PS2EXE**, on the other hand, creates an executable file from C# code that internally runs a Powershell runspace with the code of the script. **PS2EXE** has to handle input and output.
This is the reason why the two tools behave differently in borderline situations.

I do not know, why output is split over lines. My guess is that the output comes slowly because of the heavy load of the programme and therefore _PS2EXE_ assumes that after one character there is no more output and starts a new line. Can this be?

Is it an option for you to use
```powershell
Start-Process ".\qaac.exe" -ArgumentList @("-V","94",".\aud.m4a")
```
instead, maybe with the additional parameter _-NoNewWindow_ (which might create new issues on screen layout or input)?

Greetings

Markus

**Comment by mbassiouny33** [comment]
> ISESteroids creates a start-up EXE file, which then starts a hidden Powershell with the "converted" script. The hidden Powershell then behaves exactly like a normal Powershell on input and output.

That makes much more sense with your explanation!

Thanks a lot for taking the time to explain the difference.

>  and therefore PS2EXE assumes that after one character there is no more output and starts a new line. Can this be?
Probably, it is! QAAC does not output "new lines" but instead it overwrites the previously printed line.

Something close to the behavior of something like this:

```
somefor_loop{
$i # some perectenage
 Write-Host -NoNewLine "`r$i% complete 
}
```

>Is it an option for you to use

Oh yeah, I think I can live with the extra window popping up, or worst comes to worst I will redirect the output to $NULL as you suggested above, not what I initially wanted, but at least users running the script won't see random scary characters on the screen.

Thanks for your help

**Comment by mbassiouny33** [comment]
I found a workaround for this.

The idea is to launch the qaac command with CMD and redirect the output from CMD.

```
cmd /s /c --% " .\qaac.exe --verbose -V 94 .\aud.m4a" 2>&1
```

You can now print the progress on the same line like qaac does initially, or even re-write your own progress bar manually with some characters (since write-progress is not implemented in ps2exe) by extracting the progress percentage with select string

```
cmd /s /c --% " .\qaac.exe --verbose -V 94 .\aud.m4a" 2>&1 |Select-String '\[([^\[]*)\]'| ForEach-Object {
    $prog=$_.Matches.Groups[0].Value
    # $ $a do some boring string splitting on $prog to extract the percentage to int and put it in $a
    $str = "#"*$a
    $str2 = "-"*(100-$a)  
    Write-Host -NoNewLine "`r$prog complete | $str $str2|"
}
```
This idea may be useful for other external programs that behave similarly, mainly encoders, but also curl, node, etc

**Comment by MScholtes** [comment]
Hello @mbassiouny33,

it's good that you found a solution for yourself. I would not have thought of anything else.

Greetings

Markus

**Comment by mbassiouny33** [comment]
I found it randomly, initially I was looking for a way to pipe my first encoder's output (on std out) to a second software that takes bytes stream on std in. Things were not working, I found out PowerShell piping sucks and does not work like linux and CMD pipe, it tries to convert to string between the 2 programs and therefore the 2nd program does not get what it was expecting.

So I found this solution for the piping issue initially to borrow cmd to do the piping properly for me, then it hit me, what If I can use that for this issue as well to "tone down" the encoder's progress output and make CMD do the heavy lifting for me and it worked like a charm. Thought I would share it for future users as the idea is useful for anyone dealing with progressive output from node, curl, encoders(ffmpeg, qaac, x264 cli, etc) and many others tools. 

And it also shows how to compensate for the lack of "write-progress" implementation in ps2exe.

All the best!
